services/thermal_mgr/thermal_mgr.c

```diff
@@ -41,21 +43,86 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 
 }
 
+/*
+ * @brief Sends event to the thermal manager queue. Fails is queue is full.
+ *
+ * @param event - Pointer to a thermal_mgr_event_t to add to the queue.
+ */
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
-  /* Send an event to the thermal manager queue */
-
-  return ERR_CODE_SUCCESS;
+    /* Send an event to the thermal manager queue */    
+    // Check for invalid parameter
+    if( event == NULL ) {
+        return ERR_CODE_INVALID_ARG;
+    }
+    // Check if the queue has been created
+    if( thermalMgrQueueHandle == NULL ) {
+        return ERR_CODE_UNKNOWN;
+    }
+
+    switch( xQueueSend( thermalMgrQueueHandle, (void *) event, 0 ) ) {
+        case pdTRUE:
+            return ERR_CODE_SUCCESS; 
+            break;
+        case errQUEUE_FULL:
+            return ERR_CODE_QUEUE_FULL;    
+            break;
+        default:
+            return ERR_CODE_UNKNOWN;
+    }
 }
-
-void osHandlerLM75BD(void) {
-  /* Implement this function */
+/* 
+ * @brief Handles overtemperature events from temperature sensor. 
+ */
+void osHandlerLM75BD() {    
+    recomputeOTStateFlag = 1;
+    // Reset overtemperature   
+    thermal_mgr_event_t resetEvent;
+    resetEvent.type = THERMAL_MGR_EVENT_MEASURE_TEMP_CMD;  
+
+    thermalMgrSendEvent( &resetEvent );   
 }
 
-static void thermalMgr(void *pvParameters) {
-  /* Implement this task */
-  while (1) {
-    
-  }
+/*
+ * @brief Task responsible for parsing events in its queue 
+ *
+ * @param pvParameters - Configuration struct for the LM75BD temp sensor 
+ */
+static void thermalMgr(void *pvParameters) {      
+    // XXX: It is assumed we don't have to initialize temperature sensor
+    lm75bd_config_t *tempSensorConfig;  
+    error_code_t errCode;
+    tempSensorConfig = (lm75bd_config_t *) pvParameters;   
+
+    while (1) {  
+        thermal_mgr_event_t *event_recieved;
+        if ( xQueueReceive( thermalMgrQueueHandle, (void *) event_recieved, 0) == pdTRUE ) {    
+            // We got an event from the thermalMgr queue, parse it
+            switch( event_recieved->type ){  
+                case THERMAL_MGR_EVENT_MEASURE_TEMP_CMD:    
+                        // Read temperature 
+                        float temperatureResult = 0.0;   
+                        errCode = readTempLM75BD( tempSensorConfig->devAddr, &temperatureResult );  
+                       
+                        // Only change overtemperature state if we had an interrupt raised recently
+                        if (recomputeOTStateFlag) {  
+
+                            if ( temperatureResult > 80.0 ) { 
+                                overTemperatureDetected();
+                            } else {
+                                safeOperatingConditions();
+                            }  
+                            // We've handled the flag, reset it
+                            recomputeOTStateFlag = 0;
```

Comment:
Since there could be an interrupt between the if statement and when this line of code is executed, if we had an OT interrupt and then while in this if condition it returns to safe operating conditions and triggers another interrupt before this line is executed, the ISR will write a value of one (which is already the value of the flag) but this line will reset the flag to 0 anyways after that and we'll be put in a state of OT even though we have returned to safe conditions. For this reason it would be better to use something like a semaphore or queue message sent from the ISR. This approach with a counter instead of a flag to make sure you don't miss any interrupts handling might also work but is more complex solution